docs/Performance and Tuning/Workload Tuning.rst

```diff
@@ -614,9 +617,13 @@ settings must be disabled to disable AIO.
 PostgreSQL
 ~~~~~~~~~~
 
-Make separate datasets for PostgreSQL's data and WAL. Set ``recordsize=8K``
-on both to avoid expensive partial record writes. Set ``logbias=throughput``
-on PostgreSQL's data to avoid writing twice.
+Make separate datasets for PostgreSQL's data and WAL. Set 
+``compression=lz4`` and ``recordsize=32K`` (64K also work well, as 
```

Comment:
Do you have any source with benchmarks?

Larger recordsize truly can help with read-mainly workload, but on writes it adds heavily for write amplification. Default recommendations should be well-suited for all cases, even if some of them won't be optimal. Usually default can be changes until one of workload won't degrade drastically. Write amplification is such degradation.

Reply:
But it does NOT actually increase write amplification significantly, because of the compression reducing the total amount of data written. Large number of runs with pgbench with many different loads, plus our own internal loads, show significant improvement in both read and write with 16K and 32K blocks, more minor improvements at 64K and 128K, plateau at 26K and 512K, tiny (in the noise) degradation at 1M. Note that I separately timed the initial data load vs the mixed transaction pseudo-TPC portion, and the pure write phase was significantly faster.

Consider 3 cases:

- pure/heavy read load: as you pointed out, of course it helps
- pure/heavy write load: PG is writing multiple blocks at a time, so the write amplification does not come into play
- mixed load: reads are faster, single block updates are slower, multi-block updates are faster

There are a number of blog posts on ZFS tuning for PostgreSQL, and every one that has been written since LZ4 was available comes to the same conclusion.

The advice to use 8K record size was correct prior to LZ4 and provided a significant performance improvement. It is of course still better than 128K record size without compression. But LZ4 + record size larger than PG block size offers even better performance.

Reply:
LGTM.

Reply:
> But it does NOT actually increase write amplification significantly, because of the compression reducing the total amount of data written.

The key part here is amplification size, I agree it may affect less than gains, but it will raise

> pure/heavy write load: PG is writing multiple blocks at a time, so the write amplification does not come into play

Could you describe more here? It looks so for WAL, but not for data files IIUC, in case with `UPDATE`s, no?

Reply:
@shribe If you have some benchmarks' data next time - it would be great to have it in PR for history, thank you.

Reply:
> Could you describe more here? It looks so for WAL, but not for data files IIUC, in case with UPDATEs, no?

100% true for WAL, less so for tables. But WAL is about 1/2 the writes, so it's important. IIRC, I did not use separate WAL and data in my tests. Hmmm.

PG has slightly more bias toward sequential appends to end of table file than other DBs, because of the MVCC model limiting the conditions under which update-in-place can happen. But reuse of space for known "dead" tuples adds back a degree of random access.

Reply:
> But reuse of space for known "dead" tuples adds back a degree of random access.

Yep, that's most interesting part of this question for me, because usually OLTP may not grow so fast, but will change (delete/insert or just update), and eventually it will reuse space after VACUUMing, for example (IIRC).